mmengine/optim/scheduler/param_scheduler.py

```diff
@@ -1283,3 +1283,203 @@ def get_position_from_periods(
+            factor: float = 0.1,
+            patience: int = 10,
+            threshold: float = 1e-4,
+            threshold_rule: str = 'rel',
+            cooldown: int = 0,
+            min_value: float = 0,
+            eps: float = 1e-8,
+            begin: int = 0,
+            end: int = INF,
+            last_step: int = -1,
+            by_epoch: bool = True,
+            verbose: bool = False):
+
+        # Attach optimizer
+        if not isinstance(optimizer, (Optimizer, OptimWrapper)):
+            raise TypeError('``optimizer`` should be an Optimizer,'
+                            'but got {}'.format(type(optimizer).__name__))
+        self.optimizer = optimizer
+        self.param_name = param_name
+
+        if end <= begin:
+            raise ValueError('end should be larger than begin, but got'
+                             ' begin={}, end={}'.format(begin, end))
+        self.begin = begin
+        self.end = end
+
+        assert by_epoch, \
+            f'Now {type(self).__name__} only support by_epoch=True'
+        self.by_epoch = by_epoch
+
+        assert isinstance(last_step, int) and last_step >= -1
+        # Initialize valid step count and base values
+        if last_step == -1:
+            for group in optimizer.param_groups:
+                # If the param is never be scheduled, record the current value
+                # as the initial value.
+                group.setdefault(f'initial_{param_name}', group[param_name])
+        else:
+            for i, group in enumerate(optimizer.param_groups):
+                if f'initial_{param_name}' not in group:
+                    raise KeyError(
+                        f"param 'initial_{param_name}' is not specified "
+                        'in param_groups[{}] when resuming an optimizer'.
+                        format(i))
+
+        self.last_step = last_step
+
+        self._global_step = -1
+        self.verbose = verbose
+
+        if factor >= 1.0:
+            raise ValueError('Factor should be < 1.0.')
+        self.factor = factor
+
+        if isinstance(min_value, list) or isinstance(min_value, tuple):
+            if len(min_value) != len(optimizer.param_groups):
+                raise ValueError('expected {} min_lrs, got {}'.format(
+                    len(optimizer.param_groups), len(min_value)))
+            self.min_values = list(min_value)
+        else:
+            self.min_values = [min_value] * len(  # type: ignore
+                optimizer.param_groups)
+
+        self.patience = patience
+        self.cooldown = cooldown
+        self.cooldown_counter = 0
+        self.rule = rule
+        self.threshold = threshold
+        self.threshold_rule = threshold_rule
+        self.best = None
+        self.num_bad_epochs = None
+        self.rule_worse = None  # the worse value for the chosen mode
+        self.eps = eps
+
+        if monitor_stages is None:
+            self.monitor_stages = self._default_monitor_stages
```

Comment:
It seems `self.monitor_stages` could be not defined if `monitor_stages` is not None